File: sync-core/src/main/java/dev/thompgt/habitsync/sync/Change.java
package dev.thompgt.habitsync.sync;

import java.util.Map;
import java.util.Objects;
import java.util.UUID;

/**
 * One mutation to one entity — the atom that flows over the wire and through
 * {@link MergeEngine}.
 *
 * <p>A change is immutable and self-describing: everything needed to merge it is
 * present, so it can be applied on any replica, in any order, any number of times.
 *
 * @param opId       client-generated identifier, stable across retries. This is what
 *                   makes push idempotent: a request that times out after the server
 *                   committed is safe to replay, because the server recognises the
 *                   {@code opId} and treats the replay as a no-op instead of
 *                   double-applying it.
 * @param entityType which table this targets
 * @param entityId   client-generated entity id (UUIDv7 in practice, so ids sort roughly
 *                   by creation time and index locality stays reasonable)
 * @param kind       what the change does
 * @param hlc        the originating device's clock reading, and the sole basis for
 *                   conflict resolution
 * @param fields     field writes for {@link OpKind#UPSERT}; always empty for
 *                   {@link OpKind#DELETE} and {@link OpKind#RESTORE}
 */
public record Change(
        UUID opId,
        EntityType entityType,
        UUID entityId,
        OpKind kind,
        Hlc hlc,
        Map<String, FieldValue> fields) {

    public Change {
        Objects.requireNonNull(opId, "opId");
        Objects.requireNonNull(entityType, "entityType");
        Objects.requireNonNull(entityId, "entityId");
        Objects.requireNonNull(kind, "kind");
        Objects.requireNonNull(hlc, "hlc");
        fields = Map.copyOf(Objects.requireNonNull(fields, "fields"));

        if (kind != OpKind.UPSERT && !fields.isEmpty()) {
            throw new IllegalArgumentException(
                    kind + " changes must not carry field writes, got: " + fields.keySet());
        }
    }

    public static Change upsert(
            UUID opId, EntityType type, UUID entityId, Hlc hlc, Map<String, FieldValue> fields) {
        return new Change(opId, type, entityId, OpKind.UPSERT, hlc, fields);
    }

    public static Change delete(UUID opId, EntityType type, UUID entityId, Hlc hlc) {
        return new Change(opId, type, entityId, OpKind.DELETE, hlc, Map.of());
    }

    public static Change restore(UUID opId, EntityType type, UUID entityId, Hlc hlc) {
        return new Change(opId, type, entityId, OpKind.RESTORE, hlc, Map.of());
    }

    /** Identifies the entity this change targets, for grouping and lookup. */
    public EntityKey key() {
        return new EntityKey(entityType, entityId);
    }
}

File: ios/Sources/HabitSyncCore/Change.swift
import Foundation

/// One mutation to one entity — the atom that flows over the wire and through
/// ``MergeEngine``.
///
/// A change is immutable and self-describing: everything needed to merge it is present, so
/// it can be applied on any replica, in any order, any number of times.
public struct Change: Hashable, Sendable {

    /// Client-generated identifier, stable across retries.
    ///
    /// This is what makes push idempotent: a request that times out after the server
    /// committed is safe to replay, because the server recognises the op id and treats the
    /// replay as a no-op instead of double-applying it.
    public let opId: UUID

    /// Which table this targets.
    public let entityType: EntityType

    /// Client-generated entity id.
    public let entityId: UUID

    /// What the change does.
    public let kind: OpKind

    /// The originating device's clock reading, and the sole basis for conflict resolution.
    public let hlc: Hlc

    /// Field writes for ``OpKind/upsert``; always empty for `delete` and `restore`.
    public let fields: [String: FieldValue]

    public init(
        opId: UUID,
        entityType: EntityType,
        entityId: UUID,
        kind: OpKind,
        hlc: Hlc,
        fields: [String: FieldValue]
    ) {
        precondition(
            kind == .upsert || fields.isEmpty,
            "\(kind.rawValue) changes must not carry field writes, got: \(fields.keys.sorted())")
        self.opId = opId
        self.entityType = entityType
        self.entityId = entityId
        self.kind = kind
        self.hlc = hlc
        self.fields = fields
    }

    public static func upsert(
        opId: UUID = UUID(),
        type: EntityType,
        entityId: UUID,
        hlc: Hlc,
        fields: [String: FieldValue]
    ) -> Change {
        Change(opId: opId, entityType: type, entityId: entityId, kind: .upsert, hlc: hlc, fields: fields)
    }

    public static func delete(
        opId: UUID = UUID(), type: EntityType, entityId: UUID, hlc: Hlc
    ) -> Change {
        Change(opId: opId, entityType: type, entityId: entityId, kind: .delete, hlc: hlc, fields: [:])
    }

    public static func restore(
        opId: UUID = UUID(), type: EntityType, entityId: UUID, hlc: Hlc
    ) -> Change {
        Change(opId: opId, entityType: type, entityId: entityId, kind: .restore, hlc: hlc, fields: [:])
    }

    /// Identifies the entity this change targets, for grouping and lookup.
    public var key: EntityKey { EntityKey(type: entityType, id: entityId) }
}

/// A change paired with the sequence number the server assigned it.
///
/// The sequence is the server's, not the originating device's: it is the position in the
/// per-user replication log, allocated under a row lock at commit time (ADR-002). It has
/// nothing to do with the change's ``Hlc`` and must never be used to order a merge — two
/// changes can arrive in sequence order 7, 8 while their HLCs say the opposite, and the
/// HLC is the one that decides who wins.
///
/// Its only job is to be a resumable cursor: "give me everything after 42".
public struct SequencedChange: Hashable, Sendable {
    /// Position in the user's replication log, strictly increasing.
    public let serverSeq: Int64
    public let change: Change

    public init(serverSeq: Int64, change: Change) {
        precondition(serverSeq >= 1, "serverSeq must be >= 1, got \(serverSeq)")
        self.serverSeq = serverSeq
        self.change = change
    }
}
